### ISSUE TITLE
New Feature Proposal - Auto-Gen Todo List

### ISSUE THREAD

**Issue by richseviora** [eslint/eslint#9844]
Right now there isn't (to my knowledge, please correct me if I'm wrong 😀  ) the capability for ESLint to suppress the current violations so existing code bases can adopt ESLint with CI without having to fix their existing files first. 

I'd like to change that, using Rubocop's [auto-generated configuration](http://rubocop.readthedocs.io/en/latest/configuration/#automatically-generated-configuration) as inspiration. I've already developed a proof of concept solution that:

* Generates a `.eslintrc-todo.yml` based on the currently reported violations, and disables those rules for the affected files:
```yaml
overrides:
  -
    rules:
      class-methods-use-this: 0
    # Note that files are excluded on a per-rule basis.
    files:
      - src/ZhuLi/DoTheThing.js
      - src/ZhuLi/DoTheOtherThing.js
```
* Adds the new `.eslintrc-todo.yml` entry to the `extends` entry and recreates the `.eslintrc.yml` file.

# Proposed Solution

I'm planning to:

* Leverage the solution above and add support for updating the other configuration file types (with the exception of `.eslintrc.js`, not sure how I'd update a JS file). 
* Wire the solution to the CLI and add a `--create-todo` option.
* Exclude the todo list config file when re-running with `--create-todo`.

# Questions

1. Would a PR for this feature be accepted?
2. Are there any issues with the high-level approach I've outlined?
3. Are there any areas I should pay close attention to as I'm building this?

**Comment by ilyavolodin** [comment]
ESLint already has something like that built in (not exactly, but close). If you run `eslint --init` one of the options it generate config file based on your existing code. We will scan all of the files that you will pass to us and generate config that results in minimal number of errors. While this proposal is slightly different, I'm not sure the difference is big enough to justify adding a new and reasonably complicated feature.

**Comment by platinumazure** [comment]
I think this is an awesome idea, but I don't think it would fit into ESLint core.

I think you could roll your open app pretty easily. Just use our [`CLIEngine` API](https://eslint.org/docs/developer-guide/nodejs-api#CLIEngine) to run the linter and get the output, then filter through the results based on whether or not they are covered by TODOs in the external file. At that point, you could then call one of the core formatters if you wanted and echo the results to the console.

Not speaking for the team, just myself, but I would probably be interested in contributing to this if you do start a project around this idea. I've been interested in this for a while but haven't had time to do any design. If you have a reasonably good design for representing those TODOs externally, I could help with integration.

**Comment by not-an-aardvark** [comment]
Thanks for creating an issue. I think the use case described here is pretty compelling -- I've found that this is a pretty common problem when introducing a linter to a codebase. In my opinion, `eslint --init` doesn't really solve this problem because one would generally want to introduce a stricter style than what is currently present in the files.

I also like the fact that this doesn't require any changes to ESLint core, aside from generating the config file (after the config is generated, all future linting runs work the same way as they currently do). For the same reason, though, I'm wondering if this actually needs to be a part of ESLint core, as opposed to being a separate tool.

As a minor comment on the proposal: it might be better to organize the generated config file by file, rather than by rule:

```yml
overrides:
  - files: src/ZhuLi/DoTheThing.js
    rules:
      class-methods-use-this: off
      semi: off
```

I assume that people will fix problems on a file-by-file basis, and organizing it like this would make it easier to delete sections from the generated config after the problems in a given file is fixed.

**Comment by richseviora** [comment]
Thanks for the prompt responses! ❤️ 

Here's my thoughts (in no particular order):

Like @not-an-aardvark mentioned, `eslint --init` works great when you're introducing it to a codebase that's made an effort to adhere to a style guide, and you want to adhere to it. It's not as helpful when you're introducing a new style rule set to improve the quality of the existing code, but don't have the bandwidth to update all the existing files.

My opinion is that there's a good case to integrate this into ESLint core because it eases the adoption process. My client introduced Rubocop with the todo list approach, and the adoption process was very straightforward. Out of the gate, we (client's dev team and I) know that any new files we add will be properly linted, and any changes that introduce a new class of error to a file will be caught. We also make an effort to fix any to-do list issues as we touch existing files.

Compare and contrast with adopting a style guide (AirBnb as an example) to our JS project. We can't turn on ESLint for CI/CD until the entire project is fixed, and that means that we'll be potentially (almost certainly, TBH) introducing new linting issues.

@platinumazure I like the way you think! 😄 The [proof of concept](https://gist.github.com/richseviora/67595789329e938ebcbf9a4e356c50e1) leverages `CLIEngine` to generate the list of current violations, and then I do my work to transform the data, and write the updated config files.

I don't know if there's a need to represent the TODOs any differently then what we'd be using with the `overrides` logic, but I could _easily_ be missing something obvious. What are your thoughts?

I'd also love to work on this with you, either as a PR to this project or as a separate project, depending on how that decision works out.

@ilyavolodin re: your point about complicated — so far the solution hasn't proved to be particularly complicated. I think I wrote the POC up in an hour and a bit. I think the bulk of the outstanding complexity will come from dealing with different configuration file types.

@not-an-aardvark re: changing the output structure; that's a great suggestion and reflects my experience with Rubocop. When I'm remediating legacy files, I do so on a per-file basis. I think I was too locked into the Rubocop structure which keys their entries by the rule ID.

**Comment by platinumazure** [comment]
@richseviora 

> I don't know if there's a need to represent the TODOs any differently then what we'd be using with the `overrides` logic, but I could _easily_ be missing something obvious. What are your thoughts?

Actually, to be honest, the last time I really thought about this was before our glob-based configuration was added. So I wouldn't say you're missing anything-- the miss was on my part by not adapting my thinking to the current tools available :smile:

My original "design" was probably overthinking things-- I thought it would be necessary to track line numbers of existing violations, so that *only those* violations would be ignored and new violations would be flagged after the initial "create TODOs" run. And then I went down the rabbit hole about trying to reconcile as new code was added (which would throw off those line numbers), and then my head started to spin. :laughing: I think working on a per-file, per-rule basis is *probably* sufficient and we could use overrides as you suggested, since users could simply fix one file/one rule's worth of errors and then rerun the TODO tool to get that file/that rule to be removed from the overriding config, as needed.

I'll check out the POC next!

**Comment by richseviora** [comment]
I'm familiar with "probably overthinking" things too. 🤣 

The workflow I've found works best is almost what you described, the only difference is that I drop the overrides for that file first and then fix them. Lets me take advantage of IDE integration instead of having to drop back out to CLI to check as necessary.

**Comment by not-an-aardvark** [comment]
After thinking about this more, I'm wondering how effective it would be at the goal of introducing a linter while suppressing legacy warnings. Presumably, when introducing a linter, you want to ensure that new code follows the linting guidelines. But the config files that we're describing would ignore rules across an entire file, which might not only contain old code. For example, if a file has missing semicolons, then the `semi` rule will be ignored across the entire file, including new code added to the file after the config is introduced.

So this seems like it would be useful for projects that have relatively few linting errors -- for projects with a lot of errors, this might only be marginally better than adding old files to `.eslintignore`.

**Comment by platinumazure** [comment]
> Presumably, when introducing a linter, you want to ensure that new code follows the linting guidelines. But the config files that we're describing would ignore rules across an entire file, which might not only contain old code.

Right, which is why I had originally been thinking of tracking line numbers etc. But I think there might be a better way to do this: Implement the ability to lint a file (with all of the rules suppressed for the file set to error), but only report errors on lines that were changed (according to the diff between the current version and the previous version), similar to what James Henry described in our committers chat a few days ago. In that case, it might be possible for the tool to lint the file, report only errors on new/changed lines of code, and remove a rule's override when a file is completely free of errors for that rule.

**Comment by pmcelhaney** [comment]
Some other ways to approach a legacy codebase. They can all be used together, along with the above proposal. 

* This workflow has worked well for me:

  1. Set up the config the way I want it.
  2. Change any rules that don't pass to "warn". 
  3. Set `--max-warnings` to the current number of warnings for CI (and occasionally ratchet it down).
  4. Fix violations as I come across them in the IDE.
  5. Change "warn" to "error" as all the violations for each rule are fixed. 
 
*  [eslint-nibble](https://www.npmjs.com/package/eslint-nibble) came in handy. In moments where I was too tired or didn't have enough time to do "real" work, I would run `eslint-nibble` to find a rule that only had a few warnings left and fix them all at once.

* That was all before `--fix` and [prettier](https://prettier.io/) existed.

* There was a proposal (#5659) to catch comment overrides that are no longer needed. I'm not sure if anything came of it, but it strikes me as a good idea.

* For completeness, `eslint --init`.

**Comment by not-an-aardvark** [comment]
> There was a proposal (#5659) to catch comment overrides that are no longer needed. I'm not sure if anything came of it, but it strikes me as a good idea.

The `--report-unused-disable-directives` flag can be used to do this.

**Comment by richseviora** [comment]
So keeping this conversation going:

@not-an-aardvark is right; a blanket override for the entire file means that new violations on other lines won't be caught. In practice, I haven't found this to be a huge issue when I'm working on a material update to a legacy file. My workflow looks something like this: 

* I'll remove all of its overrides from the to-do file. 
* (Project w/ prettier) Run prettier to format code.
* Run auto-correct to address the low-hanging fruit (a large % of errors will be fixed here).
* Fix any outstanding issues, or suppress a specific warning based on work effort required to fix.

Is it materially better than adding ignore all the legacy files via `.eslintignore`? It's not a large difference, and if ESLint included a capability to prepopulate the `.eslintignore` then I think that'd work well enough to satisfy the user story.

@platinumazure The line-level linting handling is something I've seen in [Overcommit](https://github.com/brigade/overcommit) pre-commit hooks... and the idea sounds really interesting. I'm trying to wrap my head around the workflow and I've had a long day so I'm not having much success :)

Absent support for line-level exceptions in overrides, how would this work from a CI perspective? Or is it that we're looking to drop unused exceptions from the todo list once they no longer apply? The latter sounds like we'd need to know what the override file is so we can drop its rules from consideration during the evaluation process.

**Comment by ndushay** [comment]
I would love this feature to be added!

**Comment by richseviora** [comment]
@ndushay I've created [`eslint-takeoff`](https://www.npmjs.com/package/eslint-takeoff) as an alternative if/until this capability is added to ESLint itself. PRs, suggestions are welcome 😄

**Comment by nzakas** [comment]
Unfortunately, it looks like there wasn't enough interest from the team or community to implement this change. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to be implemented after 90 days tend to never be implemented, and as such, we [close those issues](https://eslint.org/docs/maintainer-guide/issues#when-to-close-an-issue). This doesn't mean the idea isn't interesting or useful, just that it's not something the team can commit to.
